### DISCUSSION TITLE
How to access environment variable in javascript file?

### DISCUSSION THREAD

**Discussion by andorfermichael** [nuxt/nuxt#9289] [Questions (Nuxt 2)] [answered]
I have the following files:

    // nuxt.config.js
    import { locales } from './services/i18n'
    ...
        i18n: {
            lazy: true,
            langDir: '~/locales/',
            defaultLocale: 'en',
            detectBrowserLanguage: false,
            differentDomains: true,
            locales,
            vueI18n: {
                fallbackLocale: 'en'
            }
        },
        publicRuntimeConfig: {
            ...
            subDomain: process.env.SUB_DOMAIN,
        },
    ...

--

    // services/i18n/index.js
    export const locales = [
        {
            code: 'ar',
            iso: 'ar',
            file: 'ar.json',
            dir: 'rtl',
            domain: `${process.env.SUB_DOMAIN}.example.ae`,
            name: 'العَرَبِيَّة',
            enName: 'Arabic',
            defaultLanguage: true,
            languages: ['ar']
        },
        {
            code: 'bg',
            iso: 'bg',
            file: 'bg.json',
            dir: 'ltr',
            domain: `${process.env.SUB_DOMAIN}.example.bg`,
            name: 'Български',
            enName: 'Bulgarian',
            defaultLanguage: true,
            languages: ['bg']
        },
        ...
    ]

The problem is that `process.env.SUB_DOMAIN` seems to be undefined in `/services/i18n/index.js`, although it is set because the same variable is not undefined in `nuxt.config.js`. I know that nuxt is exposing the values of `publicRuntimeConfig` as `$config`, however, `$config` is not accessible in `/services/i18n/index.js`. It might work if I would move `locales` to `nuxt.config.js`, but I don't want to do that because it would worsen the readability of the config file.

So my **question** is what the best approach would be to get subdomain inside `/services/i18n/index.js`.

**Comment by TheAlexLichter** [comment, marked as answer]
One idea that comes to my mind would be using a function instead of a plain array.


```js
// nuxt.config.js
import { locales } from './services/i18n'
...
    i18n: {
        lazy: true,
        langDir: '~/locales/',
        defaultLocale: 'en',
        detectBrowserLanguage: false,
        differentDomains: true,
        locales: locales(process.env.SUB_DOMAIN),
        vueI18n: {
            fallbackLocale: 'en'
        }
    },
    publicRuntimeConfig: {
        ...
        subDomain: process.env.SUB_DOMAIN,
    },
...
```

```js
// services/i18n/index.js
export const locales = domain => [
    {
        code: 'ar',
        iso: 'ar',
        file: 'ar.json',
        dir: 'rtl',
        domain: `${domain}.example.ae`,
        name: 'العَرَبِيَّة',
        enName: 'Arabic',
        defaultLanguage: true,
        languages: ['ar']
    },
    {
        code: 'bg',
        iso: 'bg',
        file: 'bg.json',
        dir: 'ltr',
        domain: `${domain}.example.bg`,
        name: 'Български',
        enName: 'Bulgarian',
        defaultLanguage: true,
        languages: ['bg']
    },
    ...
]
```

**Reply by kissu** [reply]
Great idea! I suggested to somehow try to access the Nuxt context into `index.js` but it is not available by any means there I guess.

**Reply by andorfermichael** [reply]
Thank you. That works for that case. However, what if I have a function in `services/i18n/index.js` that also makes use of `locales`?

For example, one of my functions looks like that?

```
export const getLocalesFromLanguage = (language, localeData) => {
    if (!localeData) localeData = locales() // same problem as before because I cannot access $config nor process.env.

    return localeData.filter((locale) => locale.iso.split('-')[0] === language)
}
```

**Reply by andorfermichael** [reply]
I also cannot rewrite this function passing `this.$i18n.locales` as `localeData` because I use this function in serverMiddleware where I cannot access `this.$i18n.locales`, if I got that right.

**Reply by andorfermichael** [reply]
ok, I simply made localeData mandatory and removed the locales call from the function.

**Comment by rafaelmagalhaes** [comment]
How would you suggest doing this with the file you just want to share env with

for example, I have a constant.js file

```javascript 
// constant.js

export const BASE_URL =  process.env.BASE_URL

```
my process.env is undefined but in nuxt.config they are all there and I can use them across plugins, middlewares, and so on
